bioamptool.py

```diff
@@ -221,104 +246,152 @@ def log_ten_minute_data(verbose=False):
 
-    # Open serial connection
-    with serial.Serial(port, baudrate, timeout=0.1) as ser:
-        csv_file = open(csv_filename, mode='w', newline='') if csv_flag else None  # Open CSV file if logging is enabled
+    try:
+        csv_file = open(csv_filename, mode='w', newline='') if csv_flag else None  # Open CSV file if logging is
         if csv_file:
             csv_writer = csv.writer(csv_file)  # Create CSV writer
             csv_writer.writerow(['Counter', 'Channel1', 'Channel2', 'Channel3', 'Channel4', 'Channel5', 'Channel6'])  # Write header
 
-        start_timer()  # Start timers for logging
+        end_time = time.time() + run_time if run_time else None
+        send_command(ser, 'START')
 
-        try:
-            ser.write(b'START\n')
-            while True:
-                read_arduino_data(ser, csv_writer)  # Read and process data from Arduino
-                current_time = time.time()  # Get the current time
-                elapsed_time = current_time - start_time  # Time elapsed since the last second
+        while True:
+            read_arduino_data(ser, csv_writer)  # Read and process data from Arduino
+            if(start_time is not None):
+                current_time = time.time()   # Get the current time
+                elapsed_time = current_time - start_time   # Time elapsed since the last second
                 elapsed_since_last_10_minutes = current_time - last_ten_minute_time  # Time elapsed since the last 10-minute interval
 
-                if elapsed_time >= 1:  # Check if one second has passed
-                    log_one_second_data(verbose)  # Log data for the last second
-                    start_time = current_time  # Reset the start time for the next second
-                if elapsed_since_last_10_minutes >= 600:  # Check if 10 minutes have passed
-                    log_ten_minute_data(verbose)  # Log data for the last 10 minutes
+                if elapsed_time >= 1:  
+                    log_one_second_data(verbose)
+                    start_time = current_time
+
+                if elapsed_since_last_10_minutes >= 600:
+                    log_ten_minute_data(verbose) 
+
                 if gui_flag:
-                    QtWidgets.QApplication.processEvents()  # Process GUI events if GUI is enabled
-                    
-                if msvcrt.kbhit() and msvcrt.getch() == b'q':  # Exit the loop if 'q' is pressed
-                    ser.write(b'STOP\n')
-                    print("Process interrupted by user")
-                    break
+                    QtWidgets.QApplication.processEvents()
 
-        except KeyboardInterrupt:
-            ser.write(b'STOP\n')
-            print("Process interrupted by user")
+                if run_time and current_time >= end_time:
+                    print("Runtime Over, sending STOP command...")
+                    send_command(ser, 'STOP')
+                    break
```

Comment:
Why do we have this under `KeyboardInterrupt`, did you test by providing end time?

Reply:
Wrong comment, resolving this now.